explorer/common/arena.h

```diff
@@ -26,32 +54,62 @@ class Arena {
+  // by this Arena.
+  //
+  // If T::EnableCanonicalizedAllocation exists and names a type, this method
+  // will canonicalize the allocated objects, meaning that two calls to this
+  // method with the same T and equal arguments will return pointers to the same
+  // object. If canonicalization is enabled, all types in Args... must be
+  // copyable, equality-comparable, and have a hash_value overload as defined in
+  // llvm/ADT/Hashing.h. If it's not possible to modify an argument type A to
+  // satisfy those requirements, the ArgKey<A> customization point can be used
+  // instead.
+  //
+  // Canonically-allocated objects must not be mutated, because those mutations
+  // would be visible to all users that happened to allocate a T object with
+  // the same constructor arguments. To help enforce this, the returned pointer
+  // will be const when canonicalization is enabled. Since that means there
+  // is no way to allocate a mutable instance of T, canonicalization should
+  // only be enabled for types that are inherently immutable.
+  //
+  // Canonicalization does not guarantee that equal objects will be identical,
+  // but it can substantially reduce the incidence of equal-but-not-identical
+  // objects, which can facilitate various optimizations.
   template <
       typename T, typename... Args,
-      typename std::enable_if_t<std::is_constructible_v<T, Args...>>* = nullptr>
-  auto New(Args&&... args) -> Nonnull<T*> {
-    auto smart_ptr =
-        std::make_unique<ArenaEntryTyped<T>>(std::forward<Args>(args)...);
-    Nonnull<T*> ptr = smart_ptr->Instance();
-    arena_.push_back(std::move(smart_ptr));
-    allocated_ += sizeof(T);
-    return ptr;
-  }
+      typename std::enable_if_t<std::is_constructible_v<T, Args...> &&
+                                !CanonicalizeAllocation<T>::value>* = nullptr>
+  auto New(Args&&... args) -> Nonnull<T*>;
+
+  template <
+      typename T, typename... Args,
+      typename std::enable_if_t<std::is_constructible_v<T, Args...> &&
+                                CanonicalizeAllocation<T>::value>* = nullptr>
+  auto New(Args&&... args) -> Nonnull<const T*>;
 
   // Allocates an object in the arena, writing its address to the given pointer.
   template <
       typename T, typename U, typename... Args,
       typename std::enable_if_t<std::is_constructible_v<T, Args...>>* = nullptr>
-  void New(WriteAddressTo<U> addr, Args&&... args) {
-    arena_.push_back(std::make_unique<ArenaEntryTyped<T>>(
-        addr, std::forward<Args>(args)...));
-    allocated_ += sizeof(T);
-  }
+  void New(WriteAddressTo<U> addr, Args&&... args);
 
   auto allocated() -> int64_t { return allocated_; }
 
  private:
+  // Allocates an object in the arena. Unlike New, this will always allocate
+  // and construct a new object.
+  template <
+      typename T, typename... Args,
+      typename std::enable_if_t<std::is_constructible_v<T, Args...>>* = nullptr>
```

Comment:
Do we need the constraint here? There's no other overload to call and we don't need this to be SFINAE-friendly.